contract_forecast/models/contract_line.py

```diff
@@ -0,0 +1,145 @@
+        )
+
+    @api.multi
+    @job(default_channel=QUEUE_CHANNEL)
+    def _generate_forecast_periods(self):
+        values = []
+        for rec in self:
+            if rec.recurring_next_date:
+                last_date_invoiced = (
+                    rec.last_date_invoiced
+                    if rec.last_date_invoiced
+                    else rec.date_start
+                )
+                period_date_end = last_date_invoiced
+                recurring_next_date = rec.recurring_next_date
+                while rec._get_generate_forecast_periods_criteria(
+                    period_date_end
+                ):
+                    period_dates = rec._get_period_to_invoice(
+                        last_date_invoiced, recurring_next_date
+                    )
+                    period_date_start, period_date_end, recurring_next_date = (
+                        period_dates
+                    )
+                    values.append(
+                        rec._prepare_contract_line_forecast_period(
+                            period_date_start,
+                            period_date_end,
+                            recurring_next_date,
+                        )
+                    )
+                    last_date_invoiced = period_date_end
+                    recurring_next_date = (
+                        recurring_next_date
+                        + self.get_relative_delta(
+                            rec.recurring_rule_type, rec.recurring_interval
+                        )
+                    )
+        return self.env["contract.line.forecast.period"].create(values)
+
+    @api.multi
+    @job(default_channel=QUEUE_CHANNEL)
+    def _unlink_forecast_periods(self):
+        return self.mapped("forecast_period_ids").unlink()
+
+    @api.model
+    def create(self, values):
+        contract_lines = super(AccountAnalyticInvoiceLine, self).create(values)
+        for contract_line in contract_lines:
+            contract_line._generate_forecast_periods()
+        return contract_lines
+
+    @api.model
+    def _get_forecast_update_trigger_fields(self):
+        return [
+            "name",
+            "sequence",
+            "product_id",
+            "date_start",
+            "date_end",
+            "quantity",
+            "price_unit",
+            "discount",
+            "recurring_invoicing_type",
+            "recurring_next_date",
+            "recurring_rule_type",
+            "recurring_interval",
+            "is_canceled",
+            "active",
+        ]
+
+    @api.multi
+    def write(self, values):
+        res = super(AccountAnalyticInvoiceLine, self).write(values)
+        if any(
+            [
+                field in values
+                for field in self._get_forecast_update_trigger_fields()
+            ]
+        ):
+            for rec in self:
+                rec._unlink_forecast_periods()
+                rec._generate_forecast_periods()
```

Comment:
@sbejaoui the unlink must be inside generate_forecast_period, to be part of the same transaction.

Also, there is a with_delay missing here, no?